---
title: "Genomics Hub opportunities"
author: Oliver Hofmann
date: "2020-12-20"
slug: hub-jobs
layout: post
permalink: https://umccr.org/blog/2020/12/20/hub-jobs
categories:
  -
tags:
  - hiring
summary: "Six job openings for bioinformaticians, data managers and curators as part of the new Illumina/UoM Genomics Hub."
---

The _tl;dr_ version: the Centre for Cancer Research is part of a new Genomics Hub, [a new collaboration between the University of Melbourne, the Victorian Government and Illumina](https://about.unimelb.edu.au/newsroom/news/2020/december/genomics-innovation-hub-to-drive-research-and-clinical-impact). As part of the initiative we are hiring:

* Two Senior Bioinformaticians (UOM 9, $123,292 - $128,275)
* Junior Bioinformatician (UOM 7, $93,935 - $101,684)
* Data Manager (UOM 7, $93,935 - $101,684)

If this sounds interesting head to the [opportunities page](https://umccr.org/work-with-us/), but if you want to have more context - read on.

### The Genomics Platform Group

Our Genomics Platform Group is part of the University of Melbourne Centre for Cancer Research ([UMCCR](https://mdhs.unimelb.edu.au/centre-for-cancer-research/home#about)). We work on workflows and methods for the rapid analysis of cancer patient's genomes and transcriptomes. Current efforts focus on WGS, WTS and ctDNA sequence analysis workflows and porting all computational resources to AWS - natively and via the new [Illumina Analytical Platform](https://mdhs.unimelb.edu.au/centre-for-cancer-research/our-research/genomics-platform-group/news-item-links/cloud-based-bioinformatics-platform-fast-tracking-genomic-analysis) (IAP). Our work is central to ongoing efforts to better understand treatment options for cancer patients through a number of different programs such as the Victorian Comprehensive Cancer Centre's [Precision Oncology program](https://www.viccompcancerctr.org/news/real-world-implementation-of-the-precision-oncology-approach/), the [Cancer of Unmet Need Initiative](https://mdhs.unimelb.edu.au/centre-for-cancer-research/our-research/precision-oncology-research-group/cancer-of-unmet-need-initiative) and our collaboration with Richard Tothill's [SUPER-NEXT](https://mdhs.unimelb.edu.au/centre-for-cancer-research/our-research/tothill-laboratory-rare-disease-oncogenomics-radio/super-next) program to tackle cancer of unknown primary patients.

### The Genomics Hub

Illumina, the University of Melbourne and Invest Victoria have launched the ["The Illumina-University of Melbourne Genomics Hub"](https://mdhs.unimelb.edu.au/centre-for-cancer-research/our-research/precision-oncology-research-group/illumina-partnership), an initiative to support genomic innovation projects and commercialization through so-called 'flagship projects'. These are supported by three core platforms -[health economics](https://mdhs.unimelb.edu.au/centre-for-cancer-research/our-research/precision-oncology-research-group/illumina-uom-partnership/platforms/health-economics-platform) (led by Maarten IJzerman), [Clinical Genomics](https://mdhs.unimelb.edu.au/centre-for-cancer-research/our-research/precision-oncology-research-group/illumina-uom-partnership/platforms/clinical-genomics-platform) (led by Sean Grimmond) and our own [bioinformatics platform](https://mdhs.unimelb.edu.au/centre-for-cancer-research/our-research/precision-oncology-research-group/illumina-uom-partnership/platforms/bioinformatics-platform). This will allow us to scale up the Cancer of Unmet Need program and expand into [non-invasive, blood based monitoring of cancer patients](https://mdhs.unimelb.edu.au/centre-for-cancer-research/our-research/precision-oncology-research-group/illumina-uom-partnership/projects/cancer-liquid-biopsy-non-invasive-blood-based-monitoring-for-cancer).

#### Bioinformaticians - Somatic Workflow Development

Our primary workflows rely on both [bcbio](https://bcbio-nextgen.readthedocs.io/en/latest/) and [DRAGEN](https://www.illumina.com/products/by-type/informatics-products/dragen-bio-it-platform.html) on AWS, but generating clinical reports from high-throughput sequencing data requires a lot of post-processing steps which we organize through the [umccrise](https://github.com/umccr/umccrise) Snakemake workflow (for WGS) and [RNAsum](https://github.com/umccr/RNAsum) (for WTS). As a senior bioinformatician you would be responsible for expanding our capabilities by:

* predicting immuno-oncological markers,
* incorporating [LINX](https://www.biorxiv.org/content/10.1101/781013v1) to better classify genomic re-arrangements.
* capturing low-allelic frequency somatic changes with dynamic thresholds,
* add short tandem repeat sensors,
* capture RNA evidence for oncovirusus,
* classifying cancers of the unknown primary,
* improving workflows for blood cancers (e.g., dealing with contaminated normal samples, or handling mixed patient/donor germlines),

.. and much more. Your work also involves improving automation on AWS and IAP, optimizing runtime through benchmarks and profiling and supervising the junior bioinformaticians on day to day production tasks, processing patient samples, analyzing accreditation data and evaluating new assays.

#### Data Manager - A National Warehouse for Cancer Genomics:

We have just received [ARDC funding](https://ardc.edu.au/news/new-data-projects-will-help-transform-australian-research/) to establish a cancer genomics warehouse in collaboration with the [Australian BioCommons](https://www.biocommons.org.au/), [ZERO Childhood Cancer](https://www.zerochildhoodcancer.org.au/), [Australian Genomics](https://www.australiangenomics.org.au/) and others. This leverages the NIH-funded [gen3 Commons](https://gen3.org/) infrastructure and aims to implement the standards and data structures required to capture, discovery and disseminate cancer data. As a Data Manager you would help drive the curation and annotation of our genomic data, coordinate with initiatives such as [ICGC-ARGO](https://www.icgc-argo.org/), ZERO and the GA4GH Cancer Working Group to harmonize clinical information, and work with our project managers to ensure all genomic information is captured in our warehouse, linked to consent and disseminated to project partners and public repositories. Your work also involves the evaluation of new approaches such as Dynamic Consent ([CTRL](https://www.australiangenomics.org.au/resources/for-patients/your-personal-platform/)) and [Federated EGA](https://ega-archive.org/federated).

#### All: Community involvement

All of our work is community-oriented - the code is open source and MIT licensed, and most of our projects are part of active collaborations with the Illumina DRAGEN team, the Hartwig Medical Foundation team and other partners. We also contribute to standards and international projects where possible, and we work closely with the [Global Alliance for Genomics and Health](https://www.ga4gh.org/). For the data manager that includes work on [DUO](https://github.com/EBISPOT/DUO), machine-readable consent and coordination with the [Genomic Knowledge Standards](https://github.com/ga4gh-gks/ga4gh-gks.github.io) workstream on data dictionaries; for the bioinformaticians this means interacting with and contributing to standards such as [TES](https://github.com/ga4gh/task-execution-schemas), [WES](https://github.com/ga4gh/workflow-execution-service-schemas), [DRS](https://ga4gh.github.io/data-repository-service-schemas/preview/release/drs-1.0.0/docs/), [htsget](https://samtools.github.io/hts-specs/htsget.html) and active participation in the workflow language communities.

If any of this sounds interesting head over to the [opportunities page](https://umccr.org/work-with-us/) or [reach out](https://umccr.org/#contact) - we would love to chat with you!
